Framework/DataHandling/src/LoadNGEM.cpp

```diff
@@ -0,0 +1,473 @@
+      if (histogramsInFrame[i].getNumberEvents() > 0) {
+        histograms[i] += histogramsInFrame[i];
+        histogramsInFrame[i].clear();
+      }
+    }
+  }
+}
+
+/**
+ * @brief Report the progress of the loader through the current file.
+ *
+ * @param numProcessedEvents The number of events processed so far.
+ * @param eventCountInFrame The number of events in the current frame.
+ * @param totalNumEvents The total number of events in the file.
+ * @param totalFilePaths The total number of file paths to process.
+ * @param fileCount The number of files processed.
+ * @return true If user has cancelled the load.
+ * @return false If the user has not cancelled.
+ */
+bool LoadNGEM::reportProgressAndCheckCancel(size_t &numProcessedEvents,
+                                            int &eventCountInFrame,
+                                            const size_t &totalNumEvents,
+                                            const size_t &totalFilePaths,
+                                            const int &fileCount) {
+  numProcessedEvents += eventCountInFrame;
+  std::string message(std::to_string(fileCount) + "/" +
+                      std::to_string(totalFilePaths));
+  progress(double(numProcessedEvents) / double(totalNumEvents) / 1.11111,
+           message);
+  eventCountInFrame = 0;
+  // Check for cancel flag.
+  return this->getCancel();
+}
+
+/**
+ * @brief Creates an event workspace and fills it with the data.
+ *
+ * @param maxToF The largest ToF seen so far.
+ * @param binWidth The width of each bin.
+ * @param histograms The main histogram event data.
+ * @param dataWorkspace The workspace to add the data to.
+ */
+void LoadNGEM::createEventWorkspace(
+    const int &maxToF, const double &binWidth,
+    std::vector<DataObjects::EventList> &histograms,
+    DataObjects::EventWorkspace_sptr &dataWorkspace) {
+  std::vector<double> xAxis;
+  for (auto i = 0; i < (maxToF / binWidth); i++) {
+    xAxis.push_back(i * binWidth);
+  }
+
+  dataWorkspace = DataObjects::create<DataObjects::EventWorkspace>(
+      NUM_OF_SPECTRA,
+      Mantid::HistogramData::Histogram(Mantid::HistogramData::BinEdges(xAxis)));
+#pragma omp parallel for
+  for (auto spectrumNo = 0u; spectrumNo < histograms.size(); ++spectrumNo) {
+    dataWorkspace->getSpectrum(spectrumNo) = histograms[spectrumNo];
+    dataWorkspace->getSpectrum(spectrumNo).setSpectrumNo(spectrumNo + 1);
+    dataWorkspace->getSpectrum(spectrumNo).setDetectorID(spectrumNo + 1);
+  }
+  dataWorkspace->setAllX(HistogramData::BinEdges{xAxis});
+  dataWorkspace->getAxis(0)->unit() =
+      Kernel::UnitFactory::Instance().create("TOF");
+  dataWorkspace->setYUnit("Counts");
+}
+
+/**
+ * @brief Create a count workspace to allow for the selection of "good"
+ * frames.
+ *
+ * @param frameEventCounts A Vector of the number of events per frame.
+ */
+void LoadNGEM::createCountWorkspace(
+    const std::vector<double> &frameEventCounts) {
+  std::vector<double> xAxisCounts;
```

Comment:
The standard algorithms are useful for going this type of thing and ensuring that we allocate the vector once and also provide a nice named call to say what we are doing:
```c++
std::vector<double> xAxisCounts(frameEventCounts.size()+1);
std::generate(xAxisCounts.begin(), xAxisCounts.end(), [n=0.0] () mutable {return ++n;});
```